core/runtime/src/main/java/io/quarkus/runtime/configuration/ConfigUtils.java

```diff
@@ -20,18 +29,116 @@ private ConfigUtils() {
+     * @return Optional value of appropriate type
+     */
+    public static Optional<?> getOptionalValue(SmallRyeConfig config, String configName, Class<?> objectType,
+            Class<? extends Converter<?>> converterClass) {
+        if (converterClass == null) {
+            return config.getOptionalValue(configName, objectType);
+        }
+
+        final Converter<?> converter = getConverterOfType(objectType, converterClass);
+        final String rawValue = config.getValue(configName, String.class);
+        return Optional.ofNullable(converter.convert(rawValue));
+    }
+
+    /**
+     * Retrieve the value of a given config name from Configuration object. Converter the value to an appropriate type using the
+     * given converter.
+     *
+     * @param config - Configuration object (must not be {@code null})
+     * @param configName - the property name (must not be {@code null})
+     * @param objectType - the type of the object (must not be {@code null})
+     * @param converterClass - The converter class to use
+     * @return the values in appropriate type
+     */
+    public static ArrayList getValues(SmallRyeConfig config, String configName, Class<?> objectType,
+            Class<? extends Converter<?>> converterClass) {
+        if (converterClass == null) {
+            return config.getValues(configName, objectType, ArrayListFactory.getInstance());
+        }
+
+        final Converter<?> converter = getConverterOfType(objectType, converterClass);
+        final ArrayList<String> rawValues = config.getValues(configName, String.class, ArrayListFactory.getInstance());
+        return rawValues.parallelStream().map(converter::convert).collect(Collectors.toCollection(ArrayList::new));
+    }
+
+    /**
+     * Converter the value to an appropriate type using the given converter.
+     *
+     * @param config - Configuration object (must not be {@code null})
+     * @param value - the value (must not be {@code null})
+     * @param objectType - the type of the object (must not be {@code null})
+     * @param converterClass - The converter class to use
+     * @return the value
+     */
+    public static Object convert(SmallRyeConfig config, String value, Class<?> objectType,
+            Class<? extends Converter<?>> converterClass) {
+        if (converterClass == null) {
+            return config.convert(value, objectType);
+        }
+
+        final Converter<?> converter = getConverterOfType(objectType, converterClass);
+        final String rawValue = config.convert(value, String.class);
+        return converter.convert(rawValue);
+    }
+
+    private static Converter<?> getConverterOfType(Class<?> type, Class<? extends Converter<?>> converterClass) {
+        if (HyphenateEnumConverter.class.equals(converterClass)) {
+            return new HyphenateEnumConverter((Class<? extends Enum>) type);
+        }
+
+        try {
+            return converterClass.newInstance();
```

Comment:
> I'm not really happy about having to use reflection at runtime. 

It's among the points that I wanted to see with you to get better idea than using reflection at runtime. So thanks for spotting this too.

> Is there a way that we could maybe populate a map of class to converter instance during static init? 

Will having a static `ConfigUtils#setConverterFor(type, converterClass)` called from `ConfigDefinition#loadEnhancedConverter()` be a reasonable approach? . The method will then cache the converter in a map .

Reply:
Since we need the converters to all be instantiated during static init, we'd have to store them somewhere during `io.quarkus.deployment.steps.ConfigurationSetup#finalizeConfigLoader`.  Lazily caching won't be good enough for this case.

Also we have to be careful only to cache converters that are used for config items that are visible at run time (`ConfigPhase#RUN_TIME` and `ConfigPhase#BUILD_AND_RUN_TIME_FIXED`).

Reply:
> Since we need the converters to all be instantiated during static init, we'd have to store them somewhere during `io.quarkus.deployment.steps.ConfigurationSetup#finalizeConfigLoader`. Lazily caching won't be good enough for this case.

Thanks for the pointer. I'll check the method. 

> 
> Also we have to be careful only to cache converters that are used for config items that are visible at run time (`ConfigPhase#RUN_TIME` and `ConfigPhase#BUILD_AND_RUN_TIME_FIXED`).

+1

Reply:
The strategy opted is to cache runtime explicit converters to get rid of runtime instance creation using reflection. 

The code code is untouched for build time explicit converters as they are not cached.

@dmlloyd WDYT?